GitHub Actions workflow for Linux and Windows machines

mickeygousset opened the pull request:
Edit: added @tspascoal as a co-author on this. 

@tspascoal and I have created a GitHub Actions workflow that will build express on both Linux and Windows machines.  I have not opened an issue to link to this, in order to allow all discussion to happen on the PR. I'm happy to open an issue if someone would like me to.

Average build time for Travis (Linux) and AppVeyor(Windows) pipelines currently in use is around 30 minutes.  With this workflow, I was able to build both Linux and Windows in just over 5 minutes, total. This is in part due to the number of concurrent pipelines that GitHub gives to open source projects.

If you look at the file, you'll notice I am using the same steps for both the Linux builds and the Windows builds.  This allows for code reuse, and also makes it where, if you need to modify a step, you only have to modify it in one place.

I'm making use of a matrix strategy as well.  Right now, the workflow is set to only use Linux and Windows. If we wanted to also build on Mac (which is an option), all we would need to do is add `macos-latest` to the matrix. The same with any other specific versions of Node we would want to test against.

[Here is an example of the build output.](https://github.com/mickeygousset/express/actions/runs/86911642)

[Here is an example of a specific job (Node 2.5 running on Linux)](https://github.com/mickeygousset/express/runs/615992721?check_suite_focus=true)

You can click into the jobs and then into each individual step on a job and see details of the run.

At this point, this workflow does everything the current Travis and Appveyor workflows do.  In addition, it goes a step beyond the AppVeyor workflow, because the GH Action workflow supports pulling down the nightly build for the Windows build (similar to what the Travis build does), which the current Appveyor workflow does not do.

mickeygousset committed:
Initial version of GH Action Workflow

This is the initial version of the GH Action workflow for building expressjs/express on Linux and Windows

jonchurch commented:
Related https://github.com/expressjs/discussions/issues/88

wesleytodd commented on .github/workflows/express-build.yml:
```diff
@@ -0,0 +1,110 @@
+# GitHub Action Workflow for building on Linux and Windows
+
+name: ExpressJS Express Build
```
```suggestion
name: Test
```

This whole thing is just the express tests, lets just rename the file and the title.

mickeygousset replied:
I had not plan with the naming.  Let's change it.

mickeygousset replied:
I've made this change (changing the file name)

wesleytodd commented on .github/workflows/express-build.yml:
```diff
@@ -0,0 +1,110 @@
+
+name: ExpressJS Express Build
+
+on:
```
We will want to update this to run on PRs and our major release lines I think.

dougwilson replied:
Yea, it should run on all branch pushes and on prs

mickeygousset replied:
Easily enough done.  We can modify the triggers to be whatever you would like.  I'll add in a trigger for PRs to master.

dougwilson replied:
It should do prs to any branch. For example we make prs all the time to the release branches.

mickeygousset replied:
I've made this change.

wesleytodd commented on .github/workflows/express-build.yml:
```diff
@@ -0,0 +1,110 @@
+        experimental: [false]
+        node-mirror: ["https://nodejs.org/dist/"]
+        include:
+          - node-version: 13
```
We should just be testing on `13` as a normal release line.  Also, seems like we should add 14 and remove the nightly stuff for the normal tests.

dougwilson replied:
These switch overs in this repo have always been part of the minor release line changes. I would scope this actions pr to just changing what the current ci does over the actions and other changes like this can be separate, landing before or after this.

wesleytodd replied:
Makes sense!  Similar to the `.npmrc` comment, just wanted to state it since it was here.

mickeygousset replied:
Just to restate so I understand correctly:  you want to add 13 and 14 to the regular matrix, and remove the nightly build stuff?

mickeygousset replied:
I've made this change

dougwilson replied:
Sorry, I missed your question. No, please do not update the tested versions in this PR.

mickeygousset replied:
I'll get it fixed.

wesleytodd commented on .github/workflows/express-build.yml:
```diff
@@ -0,0 +1,110 @@
+    steps:
+    - uses: actions/checkout@v2
+      
+    - name: Set up node using nvm
```
Why are you opting for this as opposed to the `setup-node` action provided by the Actions team?

mickeygousset replied:
Because the setup-node action provided by the actions team has the same problem that the Node Tool installer task on Azure Pipelines has:  It doesn't support Node 0.10 - 3.3.  If you try and use the setup-node action on those versions, it fails.  This action used nvm to install the appropriate node version and works over all node versions that I tested.

dougwilson replied:
Good news is we are dropping all those versions in 5.0. So if that is the only reason, perhaps we target this pr to the 5.0 branch and forget about those versions.

That is just a thought. We could keep this here and then once merged into 5.0 revert it to the action.

wesleytodd replied:
Oh!  Yeah I did run into that before, we need to get them to fix that.  

@bdougieYO @adrianmg @chrisrpatterson Who do we need to contact to get these node versions supported by the setup action?

wesleytodd replied:
Edit: @chrispat Sorry I tagged the wrong handle.
Twice: @bdougie

wesleytodd replied:
While I agree it is good we will be dropping them, we still need the tests to run for atleast a year, and if they cannot add support it would block us from moving for a year.  Not optimal IMO.

dougwilson replied:
Yea, I was thinking about that after my comment. It would be nice to have this on the 4.x line.

bdougie replied:
> @bdougieYO @adrianmg @chrisrpatterson Who do we need to contact to get these node versions supported by the setup action?

👋🏾 @chrispat is the correct person to get eyes on it. 

In regards to getting nvm working with setup-node, Let me know if this thread is helpful: https://github.com/actions/setup-node/issues/32.

wesleytodd replied:
Never sure who to tag, just tagged all the GH folks I have chatted about actions with.  I will direct these questions at @chrispat in the future, thanks!

mickeygousset replied:
I'll also reach out internally to @chrispat.  [When I brought this up to previously with the Azure Pipelines task](https://github.com/microsoft/azure-pipelines-tasks/issues/8987), it got closed unfixed, saying they weren't going to support iojs at this time.

mickeygousset replied:
regardless, I suspect if they do fix it, it might be a while out. We could use the current install tool I am using until they fix it, then switch. That would allow you to move forward and target older lines.

wesleytodd replied:
Yeah, if they continue to say no to supporting what we support on Express, I am good with using this alternate approach.  That said, we probably want to chat with @dcodeIO about their support here and make sure they will be amenable to our needs.

wesleytodd replied:
Also to be clear, I would strongly support the actions team adopting this methodology (using nvm under the hood) so that the also support the official aliases we are working on.  @dcodeio is aware of this effort from our conversation here: https://github.com/pkgjs/nv/pull/5

dcodeIO replied:
Interesting. Initially created the action because I needed aliases for rolling testing and mirrors. Wasn't aware of the Node 0.10 - 3.3 problem until now. Perhaps, if you guys depend on it as well for the time being, I can add collaborators? Should probably add more tests then so we don't break each other :)

mickeygousset replied:
I was going to try and roll my own action to do this before I found @dcodeIO's action. I'm going to push to get the Actions team to update the Node tool, but I advocate for working to keep @dcodeIO  action stable for all of us, as long as it doesn't mess up any plans they might have.

wesleytodd commented on .github/workflows/express-build.yml:
```diff
@@ -0,0 +1,110 @@
+      
+    # Configure npm
+    # Skip updating shrinkwrap / lock
+    - run: npm config set shrinkwrap false
```
I have said this in other issues, but I strongly reccomend we need to remove this and just commit an `.npmrc` which disabled the lock.  This is something we should do across all libraries in all three express orgs.

dougwilson replied:
I would probably say that to keep the scope from creeping, this actions conversion should just convert what is done today into actions, and changes that that kept separate, which we can do either before or after the actions conversion.

wesleytodd replied:
I am good with that!  I just thought I would bring it up again since it had repercussions here.

mickeygousset replied:
This is where I'm gonna have to plead a little ignorance on the technology. I'm happy to do this, I just need a little hand holding or explanation.

wesleytodd commented on .github/workflows/express-build.yml:
```diff
@@ -0,0 +1,110 @@
+    - run: npm rm --silent --save-dev connect-redis
+    
+    - name: Get Major Version Number of NodeJs Using Powershell
+      run: |
```
I would be strongly in favor of preferencing `*nix`'isms over relying on powershell here.  It is much more broadly known (I know @dougwilson's experience will disagree here) but [data says otherwise](https://insights.stackoverflow.com/survey/2019#technology-_-developers-primary-operating-systems).

That said, it seems like we could remove all of this if we just restructured the test matrix.  This removes the need to have the conversation of windows vs *nix (for this part at least).

dougwilson replied:
I wouldn't disagree, lol. I was also confused why this one part was powershell and not a unix one line the rest.

wesleytodd replied:
Ok! I thought maybe because you are a windows user and in that session issue you seemed to state otherwise.  No worries :)

dougwilson replied:
So that issue was about what to document for our users using the modules, and that we should document both the linux and windows and not simply just one of them.

wesleytodd replied:
Yep! I agree there is a difference.  So my comment stands, can we get this in a cross-compatible form which does not require powershell?

mickeygousset replied:
So I answered in the comment below why I used PowerShell Core (sorry got out of order).  The main reason for using PS Core is its cross platform nature allows me to have a step that has scripting in it that can be used on all platforms.

Yes, if we restructured the matrix to only use major integer numbers (4, 5, 6, etc...) then this method isn't needed. It is strictly here to get the major Node number, so I can use that later in some of the conditional statements, to determine what to do.

mickeygousset replied:
So I "may" have a solution for this, which allows use to stick with *nix'isms, and still run cross platform.  I'll add it with the other changes I'm working on, and see if it works.

mickeygousset replied:
I've made this change.

wesleytodd commented on .github/workflows/express-build.yml:
```diff
@@ -0,0 +1,110 @@
+        npm prune
+        npm rebuild
+        }
+      shell: pwsh
```
See my above comment, I don't think we should be using powershell in this config.

mickeygousset replied:
My main reason for using PowerShell Core here was not having to duplicate steps when doing Linux vs Windows builds.  Using PowerShell Core, as this step above does, is what allows me to use this one step in both Windows and Linux (as well as Mac builds), and also allow me to specify the OS as part of the Matrix.

The other option is to have one job for running the Linux builds, with a Matrix of Node versions in it.  Those steps could be written in *nix'isms.  Then a second job for running the Windows builds, with a matrix of Node versions in it.  Those steps would be a duplicate of a lot of the steps in Linux, but where code is required, we would still have to use PowerShell Core.

Because of this, and because PowerShell Core runs on Linux/Windows/Mac, I thought take any steps that need to be cross platform and write them in PowerShell Core.

wesleytodd replied:
Am I wrong to think this can just be:

```
run: |
  npm prune
  npm rebuild
```

If there is no way to write this in a cross platform way, I would actually prefer to split it out to have a windows test run separate from the *nix ones.  Duplication is fine if it increases developer understanding, although I do agree if we can avoid it in a reasonable way that is obviously best.

wesleytodd replied:
Ok, I hadn't looked at the travis and appveyor configs recently, but I do think if it means we can have one cross compatible script it is worth removing the check for the existence of `node_modules`.  Both of these are no-ops if there is no `node_modules` so we can just unwrap them from the if.

mickeygousset replied:
@wesleytodd correct, we remove the if statement, then we can change that to what you have above.

dougwilson replied:
I believe that i added it because some of the older npms it was not a no op and they failed. Would need to validate they are no ops across the builds.

mickeygousset replied:
I can remove it and run it and see what happens :)

mickeygousset replied:
I've made this change.

wesleytodd reviewed (changes requested):
Hey! Awesome work!  One request, could you take us through some of the decisions in here?  It would be super helpful to have all of it as comments in the code for future folks.

mickeygousset committed:
Modified name of workflow to Test

Changed the name of the workflow from ExpressJS Express Build to Test, per feedback from team.

Co-Authored-By: Wes Todd <wes@wesleytodd.com>

mickeygousset committed:
Renamed YAML file to Test.yml

Renamed YAML file to Test.yml, based off feedback from the initial pull request.

mickeygousset committed:
Changed trigger to any pull request

Changed the workflow to be any pull request to any branch, per suggestions from the submitted pull request. Leaving "on push to master" in right now, to facilitate testing, and in the event someone does do a push directly to master.

mickeygousset committed:
Add Node 13 and 14 to the matrix

- Added Node versions 13 and 14 to the matrix
- Commented out nightly build from the matrix
- Added comments to YAML file

mickeygousset committed:
Modified scripts to be unix-like on all platforms

- Change to bash script that runs on all platforms, for calculating the major node versions
- removed if statement from the prune and rebuild node_modules step.
- Add more comments to the file

mickeygousset commented:
At this point I've made all the initially requested changes from the comments earlier in the pull request.

dougwilson commented on .github/workflows/Test.yml:
```diff
@@ -0,0 +1,128 @@
+      fail-fast: false
+      matrix:
+        # Added node version 13 and 14 based off feedback from the group.
+        node-version: ["0.10", "0.12", "1.8", "2.5", "3.3", "4.9", "5.12", "6.17", "7.10", "8.17", "9.11", "10.19", "11.15", "12.16", "13.0", "14.0"]
```
Sorry there may have been confusion on the thread about the versions. Please keep the versions the same so we keep this as a change to GH actions. Adding 14 is also more involved than adding it here, as we need to validate our deps first.

But it would be nice to format this as a version per line so as we keep adding and removing versions it merges better in git being separate lines.

mickeygousset replied:
Does that mean you would like me to add back the nightly builds as well?

mickeygousset replied:
@dougwilson it doesn't appear that I can break the matrix over multiple lines. It doesn't look like that is supported at this time.

dougwilson replied:
What isn't supported? Isn't this a YAML file, so just use the list notation instead of this inline notation?

dougwilson replied:
For nightly: yea, if you can. That would also double as we know how it works.

mickeygousset replied:
I'll add the nightly back.  And yes, it is a YAML, so I'm thinking I may have done it wrong. I'll give the list another shot as well.

mickeygousset replied:
Yep it was an error between the keyboard and chair. I've added the nightly builds back, and I've turned the matrix into a list.

mickeygousset replied:
And I removed versions 13 and 14 for the matrix. At this point it should be doing exactly what the travis and appveyor builds do.

mickeygousset committed:
Changed matrix to list and added nightly builds

Changed the matrix to a list format and added the nightly builds.

wesleytodd reviewed (approved):
Looks good.  Still feels a bit weird to merge this only to follow up with adding `node@14` and moving `13` off experimental, but if LGTM then do that addition.

mickeygousset commented:
So I have a "process" question. For all the conversations above in this PR, should I mark them as resolved, or should someone else?  I think I've answered all the questions asked, but didn't know the appropriate protocol.  @wesleytodd @dougwilson could you provide me some guidance?

dougwilson commented:
Hi @mickeygousset I didn't even know you had the ability to mark them as resolved 😄 . Yes, please do so! Wes and myself spoke about this issue last night and I mentioned resolving the conversations to clear it up, but if you want to do so, please go ahead. Wes gave his blessing there at the end, so I believe this is ready to land at this point. I'm targeting it into the 4.18 branch to get in the line prior to release, which will be a huge improvement to testing times!

One thing I started looking at (but never really finished) is what we need to change on our badges to get them to show the status of these GitHub Actions now vs the Travis CI + AppVeyor statues. Not sure if you have that knowledge, and I'm still going to look into it, but just wanted to mention it here in case you knew ❤️

mickeygousset commented:
I don't know if I have the ability, but it gives me a Resolve button, so I will try. :)

Yes, there are status badges. I'll figure it out and report back.

mickeygousset commented:
I have resolved everything I have a button to resolve with :)

mickeygousset commented:
GitHub Action Status Badge: https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow#adding-a-workflow-status-badge-to-your-repository

mickeygousset commented:
I think @dougwilson may still have to do something to clear up the blocking.

dougwilson commented:
Nice on the badge! I'm looking at our badge provider (badgen.net) and I _think_ the way we would do it would be to use the https://badgen.net/github#github-status-owner-repo-ref-context area, where the name of the status check that the workflow leaves would be what we would key off.

dougwilson commented:
Ok, so looking at https://github.com/badgen/badgen.net/issues/304 it seems that we could use https://badgen.net/github#github-checks-owner-repo-ref for the exact replica of that github workflow badge link through badgen.

bjohansebas commented:
hi @mickeygousset, Thank you for making this PR, this repository has now moved to GitHub Actions
